### PR TITLE
Caused friendly error to be raised in case that Model is in a models.py ...

### PR DESCRIPTION
**Pull request by jMyles**
...outside an installed app.

Ticket: https://code.djangoproject.com/ticket/24287#ticket

### CODE REVIEW

File: django/db/models/base.py [hunk 1]
```diff
@@ -114,8 +115,14 @@ def __new__(cls, name, bases, attrs):
                     app_label_index = package_components.index(MODELS_MODULE_NAME) + 1
                 except ValueError:
                     app_label_index = 1
-                kwargs = {"app_label": package_components[app_label_index]}
-
+                try:
+                    kwargs = {"app_label": package_components[app_label_index]}
+                except IndexError:
+                    raise ImproperlyConfigured((
+                    'Unable to detect the app label for model '
```
Comment (PirosB3) [inline django/db/models/base.py]: needs indentation here. run flake8

**Comment by PirosB3** [comment]
We might want to add a test here. The tricky thing is the best way of doing this ;) suggestions?

**Comment by jMyles** [comment]
@PirosB3: My second commit has a test.  It's a little strange, and will fail improperly if the args for ModelBase.__new__ ever change, but it does at least test it.

File: tests/app_loading/tests.py [hunk 2]
```diff
@@ -75,3 +78,23 @@ def test_get_models_only_returns_installed_models(self):
         self.assertNotIn(
             "NotInstalledModel",
             [m.__name__ for m in apps.get_models()])
+
+    def test_declare_model_outside_app(self):
```
Comment (PirosB3) [inline tests/app_loading/tests.py]: Sorry to be picky, could you rename the test function name to something a bit more verbose? a suggestion could be: "test_exception_raised_if_model_declared_outside_app"

File: django/db/models/base.py [hunk 3]
```diff
@@ -114,8 +115,14 @@ def __new__(cls, name, bases, attrs):
                     app_label_index = package_components.index(MODELS_MODULE_NAME) + 1
                 except ValueError:
                     app_label_index = 1
-                kwargs = {"app_label": package_components[app_label_index]}
-
+                try:
+                    kwargs = {"app_label": package_components[app_label_index]}
+                except IndexError:
+                    raise ImproperlyConfigured((
```
Comment (timgraham) [inline django/db/models/base.py]: extra parens aren't needed

File: tests/app_loading/tests.py (hunk 2 above)
Comment (jMyles) [inline tests/app_loading/tests.py]: Yes, good looking out.  Thanks.

File: django/db/models/base.py [hunk 4]
```diff
@@ -114,8 +115,14 @@ def __new__(cls, name, bases, attrs):
...
+                        'Unable to detect the app label for model '
```
Comment (timgraham) [inline django/db/models/base.py]: line breaking is a bit odd -- %s could fit on the previous line
also, one space after period

File: tests/app_loading/tests.py [hunk 5]
```diff
@@ -75,3 +78,23 @@ def test_get_models_only_returns_installed_models(self):
...
+
+        class FakeModule(models.Model):
+            __name__ = str("models_that_do_not_live_in_an_app")
+
+        sys.modules['models_not_in_app'] = FakeModule
+
+        def declare_model_outside_app():
+            models.base.ModelBase.__new__(models.base.ModelBase,
```
Comment (timgraham) [inline tests/app_loading/tests.py]: could you use hanging indent style instead?
```
__new__(
    ...,
    ...,
)
```

File: tests/app_loading/tests.py [hunk 6]
```diff
@@ -75,3 +78,23 @@ def test_get_models_only_returns_installed_models(self):
...
+                                          str('Outsider'),
+                                          (models.Model,),
+                                          {'__module__': 'models_not_in_app'})
+
```
Comment (timgraham) [inline tests/app_loading/tests.py]: Here's the style we usually use for long messages:
```
msg = (
    '....'
    '....'
)
with self.assertRaisesMessage(ImproperlyConfigured, msg):
```

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
Please go ahead and squash commits and follow the commit message guidelines when you're ready:
https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist

**Comment by jMyles** [comment]
It failed, but it looks like it was building the wrong version (67b3c5c3670af988a3d484d15af31ad78e05a2dd  instead of 09b7dbb47e9a143b968166c127b619cd4cfaa6d0)

**Comment by aaugustin** [comment]
In fact, isn't that entire block of code going away in Django 1.9?

It's just waiting for me to double check and complete the deprecation, isn't it?

**Comment by jMyles** [comment]
Hola Aymeric. :-)

Is this true?  Can I help?

**Comment by timgraham** [comment]
Yes, I think that's correct. The ticket is #21680 (notice you're under a warning that says "This will no longer be supported in Django 1.9") -- we could do this for 1.8 though if you like.

**Comment by jMyles** [comment]
Yes, now it all makes sense to me.

OK, well I guess the strange part is: msg says that it will stop being supported in 1.9, but currently (ie, in 1.7 and, if nothing happens, in 1.8) it will just raise IndexError.  So, it's really not supported currently either.  It makes sense to me to still merge this (or replace my msg with Aymeric's better one) so that, between now and 1.9, people get used to the idea that they can't have models outside an INSTALLED_APP (and don't scratch their heads over the strange IndexError).

**Comment by timgraham** [comment]
merged to 1.8 in ac576e9f454db4942335a9a5df6168d31d4f98fe, thanks.

**Pull request closed without merging**